File: tests/image-test/generate_html.py
import json
import os

def parse_color(color):
    r = int(color['r'] * 255)
    g = int(color['g'] * 255)
    b = int(color['b'] * 255)
    return f'rgb({r}, {g}, {b})'

def parse_fills(fills):
    if not fills:
        return {}
    fill = fills[0]
    if fill['type'] == 'SOLID':
        return {'background-color': parse_color(fill['color'])}
    elif fill['type'] == 'IMAGE':
        image_path = f"images/{fill['imageHash']}.png"
        return {'background-image': f'url({image_path})', 'background-size': 'cover'}
    return {}

def parse_stroke(strokes, stroke_weight):
    if not strokes:
        return {}
    stroke = strokes[0]
    if stroke['type'] == 'SOLID':
        return {'border': f'{stroke_weight}px solid {parse_color(stroke["color"])}'}
    return {}

def parse_text_style(font_name, font_size, text_align_horizontal, text_align_vertical, fills, letter_spacing, line_height):
    style = {}
    if font_name:
        style['font-family'] = font_name['family']
        if 'style' in font_name:
            style['font-weight'] = 'bold' if 'Bold' in font_name['style'] else 'normal'
    style['font-size'] = f'{font_size}px'
    style['text-align'] = text_align_horizontal.lower()
    if fills:
        style['color'] = parse_color(fills[0]['color'])
    if letter_spacing and letter_spacing['unit'] == 'PERCENT':
        style['letter-spacing'] = f'{letter_spacing["value"]}%'
    if line_height and line_height['unit'] == 'PIXELS':
        style['line-height'] = f'{line_height["value"]}px'
    elif line_height and line_height['unit'] == 'PERCENT':
        style['line-height'] = f'{line_height["value"]}%'
    return style

def generate_css(styles):
    css = ''
    for selector, props in styles.items():
        css += f'{selector} {{\n'
        for prop, value in props.items():
            css += f'  {prop}: {value};\n'
        css += '}\n\n'
    return css

def generate_html(node, styles, class_counter):
    if node['type'] == 'FRAME':
        style = {
            'position': 'absolute',
            'left': f'{node["x"]}px',
            'top': f'{node["y"]}px',
            'width': f'{node["width"]}px',
            'height': f'{node["height"]}px',
        }
        if 'layoutMode' in node:
            if node['layoutMode'] == 'VERTICAL':
                style['display'] = 'flex'
                style['flex-direction'] = 'column'
            elif node['layoutMode'] == 'HORIZONTAL':
                style['display'] = 'flex'
                style['flex-direction'] = 'row'
        if 'primaryAxisAlignItems' in node:
            style['justify-content'] = node['primaryAxisAlignItems'].lower().replace('_', '-')
        if 'counterAxisAlignItems' in node:
            style['align-items'] = node['counterAxisAlignItems'].lower().replace('_', '-')
        if 'paddingLeft' in node:
            style['padding-left'] = f'{node["paddingLeft"]}px'
        if 'paddingRight' in node:
            style['padding-right'] = f'{node["paddingRight"]}px'
        if 'paddingTop' in node:
            style['padding-top'] = f'{node["paddingTop"]}px'
        if 'paddingBottom' in node:
            style['padding-bottom'] = f'{node["paddingBottom"]}px'
        if 'itemSpacing' in node:
            style['gap'] = f'{node["itemSpacing"]}px'
        style.update(parse_fills(node.get('fills', [])))
        style.update(parse_stroke(node.get('strokes', []), node.get('strokeWeight', 0)))
        if 'cornerRadius' in node:
            style['border-radius'] = f'{node["cornerRadius"]}px'
        if 'opacity' in node:
            style['opacity'] = str(node['opacity'])

        class_name = f'.class-{class_counter[0]}'
        class_counter[0] += 1
        styles[class_name] = style

        html = f'<div class="{class_name[1:]}">'
        if 'children' in node:
            for child in node['children']:
                html += generate_html(child, styles, class_counter)
        html += '</div>'
        return html

    elif node['type'] == 'RECTANGLE':
        style = {
            'position': 'absolute',
            'left': f'{node["x"]}px',
            'top': f'{node["y"]}px',
            'width': f'{node["width"]}px',
            'height': f'{node["height"]}px',
        }
        style.update(parse_fills(node.get('fills', [])))
        style.update(parse_stroke(node.get('strokes', []), node.get('strokeWeight', 0)))
        if 'cornerRadius' in node:
            style['border-radius'] = f'{node["cornerRadius"]}px'
        if 'opacity' in node:
            style['opacity'] = str(node['opacity'])

        class_name = f'.class-{class_counter[0]}'
        class_counter[0] += 1
        styles[class_name] = style

        return f'<div class="{class_name[1:]}"></div>'

    elif node['type'] == 'TEXT':
        style = {
            'position': 'absolute',
            'left': f'{node["x"]}px',
            'top': f'{node["y"]}px',
            'width': f'{node["width"]}px',
            'height': f'{node["height"]}px',
        }
        style.update(parse_text_style(node.get('fontName'), node.get('fontSize'), node.get('textAlignHorizontal'), node.get('textAlignVertical'), node.get('fills'), node.get('letterSpacing'), node.get('lineHeight')))
        if 'textCase' in node:
            if node['textCase'] == 'UPPER':
                style['text-transform'] = 'uppercase'
        if 'textDecoration' in node:
            style['text-decoration'] = node['textDecoration'].lower()

        class_name = f'.class-{class_counter[0]}'
        class_counter[0] += 1
        styles[class_name] = style

        return f'<div class="{class_name[1:]}">{node["characters"]}</div>'

    elif node['type'] == 'VECTOR':
        # For simplicity, skip vectors or convert to SVG if possible
        return ''

    elif node['type'] == 'ELLIPSE':
        style = {
            'position': 'absolute',
            'left': f'{node["x"]}px',
            'top': f'{node["y"]}px',
            'width': f'{node["width"]}px',
            'height': f'{node["height"]}px',
            'border-radius': '50%',
        }
        style.update(parse_fills(node.get('fills', [])))
        style.update(parse_stroke(node.get('strokes', []), node.get('strokeWeight', 0)))

        class_name = f'.class-{class_counter[0]}'
        class_counter[0] += 1
        styles[class_name] = style

        return f'<div class="{class_name[1:]}"></div>'

    elif node['type'] == 'LINE':
        style = {
            'position': 'absolute',
            'left': f'{node["x"]}px',
            'top': f'{node["y"]}px',
            'width': f'{node["width"]}px',
            'height': f'{node["height"]}px',
        }
        style.update(parse_stroke(node.get('strokes', []), node.get('strokeWeight', 0)))

        class_name = f'.class-{class_counter[0]}'
        class_counter[0] += 1
        styles[class_name] = style

        return f'<div class="{class_name[1:]}"></div>'

    elif node['type'] == 'INSTANCE':
        # Treat as frame
        return generate_html({**node, 'type': 'FRAME'}, styles, class_counter)

    elif node['type'] == 'GROUP':
        # Treat as frame
        return generate_html({**node, 'type': 'FRAME'}, styles, class_counter)

    return ''

def main():
    with open('tests/image-test/prompt.txt', 'r') as f:
        content = f.read()
        # Find the JSON part
        json_start = content.find('{"id":"1:2"')
        json_data = content[json_start:]
        data = json.loads(json_data)

    styles = {}
    class_counter = [0]
    html = generate_html(data, styles, class_counter)

    css = generate_css(styles)

    with open('tests/image-test/image-test.html', 'w') as f:
        f.write(f'''<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Figma Design</title>
    <style>
{css}
    </style>
</head>
<body>
{html}
</body>
</html>''')

if __name__ == '__main__':
    main()

File: src/generateHtml.ts
import { readFileSync, writeFileSync } from 'fs';

interface Color {
  r: number;
  g: number;
  b: number;
}

interface Fill {
  type: string;
  color?: Color;
  imageHash?: string;
}

interface Stroke {
  type: string;
  color?: Color;
}

interface FontName {
  family: string;
  style?: string;
}

interface LetterSpacing {
  unit: string;
  value: number;
}

interface LineHeight {
  unit: string;
  value: number;
}

interface Node {
  type: string;
  x: number;
  y: number;
  width: number;
  height: number;
  fills?: Fill[];
  strokes?: Stroke[];
  strokeWeight?: number;
  cornerRadius?: number;
  opacity?: number;
  layoutMode?: string;
  primaryAxisAlignItems?: string;
  counterAxisAlignItems?: string;
  paddingLeft?: number;
  paddingRight?: number;
  paddingTop?: number;
  paddingBottom?: number;
  itemSpacing?: number;
  children?: Node[];
  fontName?: FontName;
  fontSize?: number;
  textAlignHorizontal?: string;
  textAlignVertical?: string;
  letterSpacing?: LetterSpacing;
  lineHeight?: LineHeight;
  textCase?: string;
  textDecoration?: string;
  characters?: string;
}

function parseColor(color: Color): string {
  const r = Math.round(color.r * 255);
  const g = Math.round(color.g * 255);
  const b = Math.round(color.b * 255);
  return `rgb(${r}, ${g}, ${b})`;
}

function parseFills(fills?: Fill[]): Record<string, string> {
  if (!fills || fills.length === 0) return {};
  const fill = fills[0];
  if (fill.type === 'SOLID' && fill.color) {
    return { 'background-color': parseColor(fill.color) };
  } else if (fill.type === 'IMAGE' && fill.imageHash) {
    const imagePath = `images/${fill.imageHash}.png`;
    return { 'background-image': `url(${imagePath})`, 'background-size': 'cover' };
  }
  return {};
}

function parseStroke(strokes?: Stroke[], strokeWeight?: number): Record<string, string> {
  if (!strokes || strokes.length === 0) return {};
  const stroke = strokes[0];
  if (stroke.type === 'SOLID' && stroke.color) {
    return { 'border': `${strokeWeight || 0}px solid ${parseColor(stroke.color)}` };
  }
  return {};
}

function parseTextStyle(
  fontName?: FontName,
  fontSize?: number,
  textAlignHorizontal?: string,
  textAlignVertical?: string,
  fills?: Fill[],
  letterSpacing?: LetterSpacing,
  lineHeight?: LineHeight
): Record<string, string> {
  const style: Record<string, string> = {};
  if (fontName) {
    style['font-family'] = fontName.family;
    if (fontName.style && fontName.style.includes('Bold')) {
      style['font-weight'] = 'bold';
    }
  }
  if (fontSize !== undefined) style['font-size'] = `${fontSize}px`;
  if (textAlignHorizontal) style['text-align'] = textAlignHorizontal.toLowerCase();
  if (fills && fills.length > 0 && fills[0].color) {
    style['color'] = parseColor(fills[0].color);
  }
  if (letterSpacing && letterSpacing.unit === 'PERCENT') {
    style['letter-spacing'] = `${letterSpacing.value}%`;
  }
  if (lineHeight) {
    if (lineHeight.unit === 'PIXELS') {
      style['line-height'] = `${lineHeight.value}px`;
    } else if (lineHeight.unit === 'PERCENT') {
      style['line-height'] = `${lineHeight.value}%`;
    }
  }
  return style;
}

function generateCSS(styles: Record<string, Record<string, string>>): string {
  let css = '';
  for (const [selector, props] of Object.entries(styles)) {
    css += `${selector} {\n`;
    for (const [prop, value] of Object.entries(props)) {
      css += `  ${prop}: ${value};\n`;
    }
    css += '}\n\n';
  }
  return css;
}

function generateHTML(node: Node, styles: Record<string, Record<string, string>>, classCounter: number[]): string {
  if (node.type === 'FRAME') {
    const style: Record<string, string> = {
      'position': 'absolute',
      'left': `${node.x}px`,
      'top': `${node.y}px`,
      'width': `${node.width}px`,
      'height': `${node.height}px`,
    };
    if (node.layoutMode === 'VERTICAL') {
      style['display'] = 'flex';
      style['flex-direction'] = 'column';
    } else if (node.layoutMode === 'HORIZONTAL') {
      style['display'] = 'flex';
      style['flex-direction'] = 'row';
    }
    if (node.primaryAxisAlignItems) {
      style['justify-content'] = node.primaryAxisAlignItems.toLowerCase().replace('_', '-');
    }
    if (node.counterAxisAlignItems) {
      style['align-items'] = node.counterAxisAlignItems.toLowerCase().replace('_', '-');
    }
    if (node.paddingLeft !== undefined) style['padding-left'] = `${node.paddingLeft}px`;
    if (node.paddingRight !== undefined) style['padding-right'] = `${node.paddingRight}px`;
    if (node.paddingTop !== undefined) style['padding-top'] = `${node.paddingTop}px`;
    if (node.paddingBottom !== undefined) style['padding-bottom'] = `${node.paddingBottom}px`;
    if (node.itemSpacing !== undefined) style['gap'] = `${node.itemSpacing}px`;
    Object.assign(style, parseFills(node.fills));
    Object.assign(style, parseStroke(node.strokes, node.strokeWeight || 0));
    if (node.cornerRadius !== undefined) style['border-radius'] = `${node.cornerRadius}px`;
    if (node.opacity !== undefined) style['opacity'] = node.opacity.toString();

    const className = `.class-${classCounter[0]}`;
    classCounter[0]++;
    styles[className] = style;

    let html = `<div class="${className.slice(1)}">`;
    if (node.children) {
      for (const child of node.children) {
        html += generateHTML(child, styles, classCounter);
      }
    }
    html += '</div>';
    return html;
  } else if (node.type === 'RECTANGLE') {
    const style: Record<string, string> = {
      'position': 'absolute',
      'left': `${node.x}px`,
      'top': `${node.y}px`,
      'width': `${node.width}px`,
      'height': `${node.height}px`,
    };
    Object.assign(style, parseFills(node.fills));
    Object.assign(style, parseStroke(node.strokes, node.strokeWeight || 0));
    if (node.cornerRadius !== undefined) style['border-radius'] = `${node.cornerRadius}px`;
    if (node.opacity !== undefined) style['opacity'] = node.opacity.toString();

    const className = `.class-${classCounter[0]}`;
    classCounter[0]++;
    styles[className] = style;

    return `<div class="${className.slice(1)}"></div>`;
  } else if (node.type === 'TEXT') {
    const style: Record<string, string> = {
      'position': 'absolute',
      'left': `${node.x}px`,
      'top': `${node.y}px`,
      'width': `${node.width}px`,
      'height': `${node.height}px`,
    };
    Object.assign(style, parseTextStyle(node.fontName, node.fontSize, node.textAlignHorizontal, node.textAlignVertical, node.fills, node.letterSpacing, node.lineHeight));
    if (node.textCase === 'UPPER') {
      style['text-transform'] = 'uppercase';
    }
    if (node.textDecoration) {
      style['text-decoration'] = node.textDecoration.toLowerCase();
    }

    const className = `.class-${classCounter[0]}`;
    classCounter[0]++;
    styles[className] = style;

    return `<div class="${className.slice(1)}">${node.characters || ''}</div>`;
  } else if (node.type === 'VECTOR') {
    // Skip for now
    return '';
  } else if (node.type === 'ELLIPSE') {
    const style: Record<string, string> = {
      'position': 'absolute',
      'left': `${node.x}px`,
      'top': `${node.y}px`,
      'width': `${node.width}px`,
      'height': `${node.height}px`,
      'border-radius': '50%',
    };
    Object.assign(style, parseFills(node.fills));
    Object.assign(style, parseStroke(node.strokes, node.strokeWeight || 0));

    const className = `.class-${classCounter[0]}`;
    classCounter[0]++;
    styles[className] = style;

    return `<div class="${className.slice(1)}"></div>`;
  } else if (node.type === 'LINE') {
    const style: Record<string, string> = {
      'position': 'absolute',
      'left': `${node.x}px`,
      'top': `${node.y}px`,
      'width': `${node.width}px`,
      'height': `${node.height}px`,
    };
    Object.assign(style, parseStroke(node.strokes, node.strokeWeight || 0));

    const className = `.class-${classCounter[0]}`;
    classCounter[0]++;
    styles[className] = style;

    return `<div class="${className.slice(1)}"></div>`;
  } else if (node.type === 'INSTANCE') {
    return generateHTML({ ...node, type: 'FRAME' }, styles, classCounter);
  } else if (node.type === 'GROUP') {
    return generateHTML({ ...node, type: 'FRAME' }, styles, classCounter);
  }
  return '';
}

function main(): void {
  const content = readFileSync('tests/image-test/prompt.txt', 'utf8');
  const jsonStart = content.indexOf('{"id":"1:2"');
  const jsonData = content.slice(jsonStart);
  const data: Node = JSON.parse(jsonData);

  const styles: Record<string, Record<string, string>> = {};
  const classCounter = [0];
  const html = generateHTML(data, styles, classCounter);

  const css = generateCSS(styles);

  const output = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Figma Design</title>
    <style>
${css}
    </style>
</head>
<body>
${html}
</body>
</html>`;

  writeFileSync('tests/image-test/image-test.html', output);
  console.log('HTML and CSS generated successfully');
}

main();